# SCOPE
**Stabilized COre gene and Pathway Election**

Analyzing gene expression data collected from paired disease-control tissues plays a critical role in characterizing disease mechanisms. Although differential expression analysis can prioritize individual genes as candidate for further investigation, joint analysis of multiple genes is yet a problem to be thoroughly tackled. Currently, state-of-the-art methods can be categorized onto two streams: (1) co-expression network analyses focusing on correlations between genes; (2) multiple linear regressions (usually regularized) to select genes jointly. Both methods suffer from the problem of stability: a slight change of parameterization or dataset could lead to dramatic alternation of outcomes, leaving uncertainties in biological interpretation. In this work, we propose Stabilized COre gene and Pathway Election, or SCOPE, a stabilized model integrating bootstrapped LASSO and co-expression analysis, leading to robust outcomes insensitive to variations in data. As a proof-of-concept, we applied SCOPE to six cancer expressions datasets in The Cancer Genome Atlas (TCGA). Our analysis identified core genes that are critical to cancers and revealed pathways shared across cancers (in contrast to the distinctive outcomes generated by standard analyses alone). As an example, we highlighted the pivotal role of CD63 as an oncogenic driver and a potential therapeutic target in kidney cancer. 

The R code for the SCOPE pipeline and instructions on how to run it are included in this repository.

## Example Scenario
Assume that the data to be analyzed consists of two expression data files from the TCGA database:
- BRCA.csv
- KIRC.csv

Assume also that the expression files are of the following form:

| Gene 1 | Gene 2 | ... | Gene k | Phenotype |
|---|---|---|---|---|
|...|...|...|...|...|

## 1_glmnet_stabilized_lasso.R
This script is used for running multiple LASSO runs and aggregating the results. The script can be run using the following command:
```bash
Rscript --max-ppsize=500000 1_glmnet_stabilized_lasso.R EXPR_FILE ITERS FOLDS START_SEED PREFIX
```
The arguments supplied here are as follows:
- EXPR_FILE - the expression data file (as shown above)
- ITERS - this denotes the numebr of LASSO models that must be trained using the given data (e.g.: ITERS=100 indicates training 100 independent LASSO models on separate training subsets of the original data) _(default = 1000)_
- FOLDS - the number of cross-validation folds to use in the lasso models _(default = 10)_
- START_SEED - random start seed _(default = 2222)_
- PREFIX - output files are prefixed using this string. Run each dataset with a different prefix such that outputs are distiguishable! _(default = "stabilized_lasso")_

**Note 1:** This script file should be modified to identify the required case and control phenotypes as well as the required filtering/quality control.

**Note 2:** It is possible to change the number of threads used by the glmnet package in the script. Default is 2. Please note that a larger thread count does not necessarily improve performance due to the overhead in copying the expression matrix.

**Note 3:** The max-ppsize=500000 is used to allow R to work with the large design matrix created in running glmnet on (typically) large expression datasets with a large number of variables (genes).

## 2_null_corr.R
This script will generate the null distribution for co-expression given the expression matrix. The script can be run using the following command:
```bash
Rscript 2_null_corr.R EXPR_FILE ITERS PROBES_PER_ITER START_SEED PREFIX
```
The arguments supplied here are as follows:
- EXPR_FILE - the expression data file (as shown above)
- ITERS - this denotes the numebr of sets of _PROBES_PER_ITER_ variables (genes/transcripts) that are selected to calculated correlation matrices _(default = 100)_
- PROBES_PER_ITER - the number of variables (genes/transcripts) that are selected in each iteration to calculate the correlation matrices _(default = 1000)_
- START_SEED - random start seed _(default = 2222)_
- PREFIX - output files are prefixed using this string. Run each dataset with a different prefix such that outputs are distiguishable! _(default = "stabilized_lasso")_

**Note:** While accuracy would typically increase with larger ITERS, larger ITERS also indicates larger files that may not fit in memory in later scripts. (In the case that a HPC is available, 1000 ITERS or higher is recommended.)

## 3_null_diff_corr.R
This script will generate the null distribution for differential co-expression given the expression matrix. The script can be run using the following command:
```bash
Rscript 3_null_diff_corr.R EXPR_FILE ITERS PROBES_PER_ITER START_SEED PREFIX
```
The arguments supplied here are as follows:
- EXPR_FILE - the expression data file (as shown above)
- ITERS - this denotes the numebr of sets of _PROBES_PER_ITER_ variables (genes/transcripts) that are selected to calculated correlation matrices. i.e. In each iteration _PROBES_PER_ITER_ number of columns are selected from the expression matrixes for both phenotypes and the differences in correlations among these calculated. _(default = 100)_
- PROBES_PER_ITER - the number of variables (genes/transcripts) that are selected in each iteration to calculate the correlation matrices _(default = 1000)_
- START_SEED - random start seed _(default = 2222)_
- PREFIX - output files are prefixed using this string. Run each dataset with a different prefix such that outputs are distiguishable! _(default = "stabilized_lasso")_

**Note:** While accuracy would typically increase with larger ITERS, larger ITERS also indicates larger files that may not fit in memory in later scripts. (In the case that a HPC is available, 1000 ITERS or higher is recommended.)

## 4_genes.R
This script will identify all secondary genes as well as the co-expressions and differential co-expressions of all genes in the dataset with each of the core genes. The script can be run using the following command:
```bash
Rscript 4_genes.R EXPR_FILE SUMMARY_FILE CORE_CUTOFF CORR_NULL_FILE CORR_PERCENTILE_THRESHOLD DIFFCORR_NULL_FILE DIFFCORR_PERCENTILE_THRESHOLD PREFIX
```
The arguments supplied here are as follows:
- EXPR_FILE - the expression data file (as shown above)
- SUMMARY_FILE - the corresponding summary file generated from [1_glmnet_stabilized_lasso.R](#1_glmnet_stabilized_lasso.R)
- CORE_CUTOFF - cutoff to identify core genes (dependent on the number of ITERS used in [1_glmnet_stabilized_lasso.R](#1_glmnet_stabilized_lasso.R)) _(default = 950)_
- CORR_NULL_FILE - co-expression null distribution generated by [2_null_corr.R](#2_null_corr.R)
- CORR_PERCENTILE_THRESHOLD - percentile threshold beyond which secondary genes are identified as co-expressed (positively *or* negatively correlated) with each core gene i.e. this percentile is used to calculate the positive and negative correlation values from each of the ITERS stored in the file and the median of these values obtained for both positive and negative values _(default = 0.975)_
- DIFFCORR_NULL_FILE - co-expression null distribution generated by [3_null_diff_corr.R](#3_null_diff_corr.R)
- DIFFCORR_PERCENTILE_THRESHOLD - percentile threshold beyond which secondary genes are identified as differentially co-expressed with each core gene i.e. this percentile is used to calculate the differential co-expression values from each of the ITERS stored in the file and the median of these values obtained _(default = 0.975)_
- PREFIX - output files are prefixed using this string. Run each dataset with a different prefix such that outputs are distiguishable! _(default = "stabilized_lasso")_

**Note:** This script may also be run using default values using:
```bash
Rscript 4_genes.R EXPR_FILE SUMMARY_FILE CORR_NULL_FILE DIFFCORR_NULL_FILE PREFIX
```

## 5_pathway_enrichment.R
This script will identify the pathways enriched by each CGN of the dataset specified. The script can be run using the following command:
```bash
Rscript 5_pathway_enrichment.R SECONDARY_GENE_FILE CANCER
```
The arguments supplied here are as follows:
- SECONDARY_GENE_FILE - the "_SecondaryGenes.csv" file generated by [4_genes.R](#4_genes.R)
- CANCER - the name of the current cancer/dataset being analyzed i.e. "BRCA" or "KIRC" in the example scenario given above

**Note:** This script may need to be modified based on the type of data being analyzed. The parameters of the [WebGestaltR](#https://cran.r-project.org/web/packages/WebGestaltR/index.html) package can be changed to specify the type of probes being used as well as the pathway database to be used.

## 6_pathway_overlaps.R
This script will identify pathways overlapped over all the datasets analyzed given in a particular directory. The script can be run using the following command:
```bash
Rscript 6_pathway_overlaps.R PATHWAY_DIR PREFIX
```
The arguments supplied here are as follows:
- PATHWAY_DIR - this directory must contain all the outputs from [5_pathway_enrichment.R](#5_pathway_enrichment.R) for each of the datasets/cancers analyzed
- PREFIX - output files are prefixed using this string. _(default = "SCOPE")_
